ARROW-12677: [Python] Add a mask argument to pyarrow.StructArray.from_arrays

westonpace opened the pull request:
This allows the user to supply an optional `mask` when creating a struct array.

 * The mask requirements are pretty strict (must be a boolean arrow array without nulls) compared with some of the other functions (e.g. `array.mask` accepts a wide variety of inputs).  I think this should be ok since this use case is probably rarer and there are other plenty of existing ways to convert other datatypes to an arrow array.
 * ~~Unfortunately, StructArray::Make interprets the "null buffer" as more of a validity buffer (1 = valid, 0 = null).  This is the opposite of everywhere else a `mask` is used.  I was torn between inverting the input buffer to mimic the python API and passing through directly to the C interface for simplicity.  I chose the simpler option but could be convinced otherwise.~~ Per request, I now invert the mask to align with the python API.

jorisvandenbossche commented:
If we add such a `mask` keyword, I would say that it should match the `mask` keyword of `pa.array(..)` (and thus being an inverted mask, and not the validity mask as used internally). 

That means always some conversion of the data is needed (to invert the mask), and you cannot create a StructArray 100% cheaply from existing arrays with `from_arrays`, but for such a case you can still use `from_buffers` if you want.

westonpace commented:
I've added the call to invert.  I went ahead and added a `memory_pool` parameter per @&res suggestion on the JIRA.  I also verified that we can create null elements in a `ListArray` and added a test for `ListArray.from_arrays` since there was none.

I believe I have addressed all PR comments and, pending CI, this is ready for review.

lidavidm commented on python/pyarrow/array.pxi:
```diff
@@ -2189,6 +2227,18 @@ cdef class StructArray(Array):
+        if mask is None:
+            c_mask = shared_ptr[CBuffer]()
+        elif isinstance(mask, Array):
+            if mask.type != bool_():
```
nit: maybe pa.types.is_boolean?

westonpace replied:
I couldn't figure out how to access `is_boolean` from the `pxi` file but I changed it to `mask.type.id != Type_BOOL` which is similar to the way other type comparisons are done in this file.

lidavidm commented on python/pyarrow/array.pxi:
```diff
@@ -2189,6 +2227,18 @@ cdef class StructArray(Array):
+            inverted_mask = _pc().invert(mask, memory_pool=memory_pool)
+            c_mask = pyarrow_unwrap_buffer(inverted_mask.buffers()[1])
+        else:
+            raise ValueError('Mask must be a pyarray.Array of type bool')
```
```suggestion
            raise ValueError('Mask must be a pyarrow.Array of type bool')
```

lidavidm replied:
nit: maybe also include an `'(expected pyarrow.Array of type bool, got {type(mask)})'` (this is semi-consistently done in PyArrow)

lidavidm commented on python/pyarrow/array.pxi:
```diff
@@ -2189,6 +2227,18 @@ cdef class StructArray(Array):
+            c_mask = shared_ptr[CBuffer]()
+        elif isinstance(mask, Array):
+            if mask.type != bool_():
+                raise ValueError('Mask must be a pyarray.Array of type bool')
```
```suggestion
                raise ValueError('Mask must be a pyarrow.Array of type bool')
```

lidavidm commented on python/pyarrow/tests/test_array.py:
```diff
@@ -932,6 +954,24 @@ def test_fixed_size_list_from_arrays():
+
+    with pytest.raises(ValueError):
+        pa.ListArray.from_arrays(pa.array([0, 2, 5]), values)
+
```
```suggestion


```

lidavidm replied:
(just to fix the lint error)

lidavidm reviewed (commented):
LGTM, thanks! Left a few nits about typos/exception messages.

lidavidm reviewed (approved):
It looks like the Python linter is unhappy, but otherwise LGTM.

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -2153,7 +2186,8 @@ cdef class StructArray(Array):
     @staticmethod
-    def from_arrays(arrays, names=None, fields=None):
+    def from_arrays(arrays, names=None, fields=None, mask=None,
+                    memory_pool=None):
```
It seems we are also inconsistent in the naming of this keyword, the `ListArray.from_arrays` above uses a `pool` keyword (but `memory_pool` is used more often, so this change is fine, will open a JIRA to make this consistent)

jorisvandenbossche replied:
-> https://issues.apache.org/jira/browse/ARROW-12805
